{% extends 'polls/base.html' %}

{% block content %}
<div class="page-header" id="banner">
    <div class="row">
    <ul class="nav nav-tabs">
      <li class="active"><a aria-expanded="true" href="#intro" data-toggle="tab">Introduction</a></li>
      <li class=""><a aria-expanded="false" href="#files" data-toggle="tab">VCF files</a></li>
    </ul>
    
<div id="myTabContent" class="tab-content">
  <div class="tab-pane fade active in" id="intro">
    <br>
    <p class="prg-menu"><i>Web App to export data from VCF files, there are some standard datasets like ESP, ExAC and 1000GP.</i></p>

    <p class="prg-menu"><i>You may also upload your own VCF file and export data from it to a Excel file, fasta file or check stats. Depending on wich dataset you choose there are different types of file outputs and filtering options.</i></p>
    
  </div>
  <div class="tab-pane fade" id="files">
    <br>
    <p class="prg-menu"><i>Please validate your VCF file before running the tool.</i></p>
    
    <p class="prg-menu"><i>VCF files have meta-information lines and a header line that has eight fixed fields including chromosome (CHROM), position (POS), known variant identifications (ID), reference allele (REF), alternate allele(s) (ALT), variant quality score (QUAL), filter information summarizing why a variant was or was not considered valid by the variant calling software (FILTER), and an information field (INFO). Additional fields containing genotypes for one or more samples may also be present. In our VCF files we have the FORMAT column and multiple sample columns.</i></p>

    <p class="prg-menu"><i>Each row of the file contains information about observed variants at the given position and chromosome. Files can contain information for a single or multiple samples corresponding to one or more columns, each column with double information (genotype) for autosomal chromosomes.</i></p>
  </div>
</div>


{% endblock %}
